src/Microsoft.ML.FastTree/TreeEnsembleFeaturizationEstimator.cs

```diff
@@ -0,0 +1,464 @@
+            /// and the i-th vector element is the prediction value predicted by the i-th tree.
+            /// If <see cref="TreesColumnName"/> is <see langword="null"/>, this output column may not be generated.
+            /// </summary>
+            public string TreesColumnName;
+
+            /// <summary>
+            /// The 0-1 encoding of all leaf nodes' IDs. Its type is a vector of <see cref="System.Single"/>. If the given feature
+            /// vector falls into the first leaf of the first tree, the first element in the 0-1 encoding would be 1.
+            /// If <see cref="LeavesColumnName"/> is <see langword="null"/>, this output column may not be generated.
+            /// </summary>
+            public string LeavesColumnName;
+
+            /// <summary>
+            /// The 0-1 encoding of the paths to the leaves. If the path to the first tree's leaf is node 1 (2nd node in the first tree),
+            /// node 3 (4th node in the first tree), and node 5 (6th node in the first tree), the 2nd, 4th, and 6th element in that encoding
+            /// would be 1.
+            /// If <see cref="PathsColumnName"/> is <see langword="null"/>, this output column may not be generated.
+            /// </summary>
+            public string PathsColumnName;
+        };
+
+        /// <summary>
+        /// See <see cref="CommonOptions.InputColumnName"/>.
+        /// </summary>
+        private protected readonly string FeatureColumnName;
+
+        /// <summary>
+        /// See <see cref="CommonOptions.TreesColumnName"/>.
+        /// </summary>
+        private protected readonly string TreesColumnName;
+
+        /// <summary>
+        /// See <see cref="CommonOptions.LeavesColumnName"/>.
+        /// </summary>
+        private protected readonly string LeavesColumnName;
+
+        /// <summary>
+        /// See <see cref="CommonOptions.PathsColumnName"/>.
+        /// </summary>
+        private protected readonly string PathsColumnName;
+
+        /// <summary>
+        /// Environment of this instance. It controls error throwing and other environment settings.
+        /// </summary>
+        private protected readonly IHostEnvironment Env;
+
+        private protected TreeEnsembleFeaturizationEstimatorBase(IHostEnvironment env, CommonOptions options)
+        {
+            Env = env;
+            if (options.InputColumnName == null)
+                throw Env.Except(nameof(options), "The " + nameof(options.InputColumnName) + " cannot be null.");
+            if (options.TreesColumnName == null && options.LeavesColumnName == null && options.PathsColumnName == null)
+                throw Env.Except($"{nameof(CommonOptions.TreesColumnName)}, {nameof(CommonOptions.LeavesColumnName)}, and {nameof(CommonOptions.PathsColumnName)} cannot be all null at the same time. " +
+                    $"At least one output column name should be provided so that at least one output column may be generated.");
```

Comment:
>that [](start = 76, length = 4)

This `that` looks out of place. Could be wrong

Reply:
Removed.

---
In reply to: [291367319](https://github.com/dotnet/machinelearning/pull/3812#discussion_r291367319) [](ancestors = 291367319)